RedHatInsights/rhsm-subscriptions
# Swatch-444: Replace MetricId enum with class that checks against product-config

Jira issue: [SWATCH-444](https://issues.redhat.com/browse/SWATCH-444)

## Description

Rather than have using an Enum for metricId we now have a class that will validate a metric ID based on what is inside swatch-product-config. This allows for easier maintainability.

## Testing


### Setup

1. Start app with: `DEV_MODE=true ./gradlew :bootRun`


### Verification

1. Verify below curl requests have no errors:
```
curl -X 'GET' \
  'http://localhost:8000/api/rhsm-subscriptions/v1/tally/products/OpenShift%20Container%20Platform/Cores?granularity=Daily&beginning=2023-07-21T17%3A32%3A28Z&ending=2023-08-21T17%3A32%3A28Z&use_running_totals_format=false' \
  -H 'accept: application/vnd.api+json' \
  -H 'x-rh-identity: eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6ImFjY291bnQxMjMiLCJ0eXBlIjoiVXNlciIsInVzZXIiOnsiaXNfb3JnX2FkbWluIjp0cnVlfSwiaW50ZXJuYWwiOnsib3JnX2lkIjoib3JnMTIzIn19fQo='
  
  curl -X 'GET' \
  'http://localhost:8000/api/rhsm-subscriptions/v1/capacity/products/OpenShift%20Container%20Platform/Sockets?granularity=Daily&beginning=2023-07-21T17%3A32%3A28Z&ending=2023-07-22T17%3A32%3A28Z' \
  -H 'accept: application/vnd.api+json' \
  -H 'x-rh-identity: eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6ImFjY291bnQxMjMiLCJ0eXBlIjoiVXNlciIsInVzZXIiOnsiaXNfb3JnX2FkbWluIjp0cnVlfSwiaW50ZXJuYWwiOnsib3JnX2lkIjoib3JnMTIzIn19fQo='
  
  curl -X 'GET' \
  'http://localhost:8000/api/rhsm-subscriptions/v1/instances/products/OpenShift%20Container%20Platform?uom=Cores' \
  -H 'accept: application/json' \
  -H 'x-rh-identity: eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6ImFjY291bnQxMjMiLCJ0eXBlIjoiVXNlciIsInVzZXIiOnsiaXNfb3JnX2FkbWluIjp0cnVlfSwiaW50ZXJuYWwiOnsib3JnX2lkIjoib3JnMTIzIn19fQo='
```

1. Verify below curl requests return 400 error:
```
curl -X 'GET' \
  'http://localhost:8000/api/rhsm-subscriptions/v1/tally/products/OpenShift%20Container%20Platform/NotAMetric?granularity=Daily&beginning=2023-07-21T17%3A32%3A28Z&ending=2023-08-21T17%3A32%3A28Z&use_running_totals_format=false' \
  -H 'accept: application/vnd.api+json' \
  -H 'x-rh-identity: eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6ImFjY291bnQxMjMiLCJ0eXBlIjoiVXNlciIsInVzZXIiOnsiaXNfb3JnX2FkbWluIjp0cnVlfSwiaW50ZXJuYWwiOnsib3JnX2lkIjoib3JnMTIzIn19fQo='
  
  curl -X 'GET' \
  'http://localhost:8000/api/rhsm-subscriptions/v1/capacity/products/OpenShift%20Container%20Platform/NotAMetric?granularity=Daily&beginning=2023-07-21T17%3A32%3A28Z&ending=2023-07-22T17%3A32%3A28Z' \
  -H 'accept: application/vnd.api+json' \
  -H 'x-rh-identity: eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6ImFjY291bnQxMjMiLCJ0eXBlIjoiVXNlciIsInVzZXIiOnsiaXNfb3JnX2FkbWluIjp0cnVlfSwiaW50ZXJuYWwiOnsib3JnX2lkIjoib3JnMTIzIn19fQo='
  
  curl -X 'GET' \
  'http://localhost:8000/api/rhsm-subscriptions/v1/instances/products/OpenShift%20Container%20Platform?uom=NotAMetric' \
  -H 'accept: application/json' \
  -H 'x-rh-identity: eyJpZGVudGl0eSI6eyJhY2NvdW50X251bWJlciI6ImFjY291bnQxMjMiLCJ0eXBlIjoiVXNlciIsInVzZXIiOnsiaXNfb3JnX2FkbWluIjp0cnVlfSwiaW50ZXJuYWwiOnsib3JnX2lkIjoib3JnMTIzIn19fQo='
```
